core/utils/codel.h

```diff
@@ -0,0 +1,331 @@
+#ifndef BESS_UTILS_CODEL_H_
+#define BESS_UTILS_CODEL_H_
+
+#include <math.h>
+#include <sys/time.h>
+
+#include <iostream>
+
+#include "./lock_free_queue.h"
+#include "./queue.h"
+
+namespace bess {
+namespace utils {
+/**
+ * Codel(Controlled Delay Management) is an Queue controller based on this
+ * article http://queue.acm.org/detail.cfm?id=2209336
+ *
+ * It provides an active queue management to help prevent bufferbloat by dropping
+ * queue entries at an increasing rate if the delay in the queue is above the
+ * target queue delay. The equation used to calculate drop intervals is based on TCP
+ * throughput response to drop probability.
+ *
+ * Takes Template argument Q which is a templated child class of Queue and T the type
+ * that is going to be enqueued/dequeued.
+ */
+template <template <typename type> class Q, typename T>
+class Codel : public Queue<T> {
+ public:
+  static constexpr uint64_t kDefaultTarget = 5000;  // default delay target for codel
+  static constexpr uint64_t kDefaultWindow = 100000;  // default window size for codel
+  static const int kDefaultSlots = 4096;  // default number of slots in the codel queue
+
+  // A wrapper class around the pointer to track the delay time and whether
+  // the packet can be dropped after dequeueing.
+  struct Wrapper {
+    T* ptr;              // a pointer to be enqueued
+    uint64_t timestamp;  // a timestamp of the time the pointer was enqueued
+    uint8_t drop;  // used during dequeue to determin if the packet can be dropped
+    Wrapper(T* obj, uint64_t stamp) : ptr(obj), timestamp(stamp), drop(0){ }
+  };
+
+  Codel(void (*drop_func)(T*), Q<Wrapper>* init_queue,
+        uint64_t target = kDefaultTarget, uint64_t window = kDefaultWindow);
+  ~Codel();
+
+  /**
+   * Enqueue one object. Takes a pointer to the object to be added and integer
+   * pointer to set on error. error codes: -1 is Quota exceeded. The objects
+   * have been enqueued, but the high water mark is exceeded. -2 is not enough
+   * room in the ring to enqueue; no object is enqueued.
+  */
+  void Push(T* ptr, int* err);
```

Comment:
Ah, yes -- I know this is probably seemingly conflicting advice.  But here's the general rules I try to follow:
* If the method has no return value, and either fails or succeeds, then return a bool indicating this (e.g. true upon success).
* If the method is as above but already has a return value, then try to include the error as part of the return (e.g. return nullptr to signal failure if the return value is a pointer).
* If the method has a more complex set of error conditions then use a int * argument (as you did here).

It seemed to me that you don't have that many error conditions, and I was wondering if you could just return a failure/success signal as a bool or as part of an existing return value (here and in other methods in this PR).